cmd/promtool/unittest.go

```diff
@@ -0,0 +1,489 @@
+						Labels:      labels.FromMap(a.ExpLabels),
+						Annotations: labels.FromMap(a.ExpAnnotations),
+					})
+				}
+
+				if gotAlerts.Len() != expAlerts.Len() {
+					errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+						testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+				} else {
+					sort.Sort(gotAlerts)
+					sort.Sort(expAlerts)
+
+					if !reflect.DeepEqual(expAlerts, gotAlerts) {
+						errs = append(errs, fmt.Errorf("    alertname:%s, time:%s, \n        exp:%#v, \n        got:%#v",
+							testcase.Alertname, testcase.EvalTime.String(), expAlerts.String(), gotAlerts.String()))
+					}
+				}
+			}
+
+			curr++
+		}
+	}
+
+	// Checking promql expressions.
+	for _, testCase := range tg.PromqlExprTests {
+		got, err := query(suite.Context(), testCase.Expr, mint.Add(testCase.EvalTime),
+			suite.QueryEngine(), suite.Queryable())
+		if err != nil {
+			errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+				testCase.EvalTime.String(), err.Error()))
+		}
+
+		var gotSamples []parsedSample
+		for _, s := range got {
+			gotSamples = append(gotSamples, parsedSample{
+				Labels: s.Metric.Copy(),
+				Value:  s.V,
+			})
+		}
+
+		var expSamples []parsedSample
+		parseError := false
+		for _, s := range testCase.ExpSamples {
+			lb, err := promql.ParseMetric(s.Labels)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, err:%s", testCase.Expr,
+					testCase.EvalTime.String(), err.Error()))
+				parseError = true
+				break
+			}
+			expSamples = append(expSamples, parsedSample{
+				Labels: lb,
+				Value:  s.Value,
+			})
+		}
+
+		if !parseError && !reflect.DeepEqual(expSamples, gotSamples) {
+			errs = append(errs, fmt.Errorf("    expr:'%s', time:%s, \n        exp:%#v, \n        got:%#v", testCase.Expr,
+				testCase.EvalTime.String(), parsedSamplesString(expSamples), parsedSamplesString(gotSamples)))
+		}
+	}
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// seriesLoadingString returns the input series in
+// PromQL notation.
+func (tg *testGroup) seriesLoadingString() string {
+	result := ""
+	result += "load " + shortDur(tg.Interval) + "\n"
+	for _, is := range tg.InputSeries {
+		result += "  " + is.Series + " " + is.Values + "\n"
+	}
+	return result
+}
+
+// https://stackoverflow.com/a/41336257/6219247
+func shortDur(d time.Duration) string {
+	s := d.String()
+	if strings.HasSuffix(s, "m0s") {
+		s = s[:len(s)-2]
+	}
+	if strings.HasSuffix(s, "h0m") {
+		s = s[:len(s)-2]
+	}
+	return s
+}
+
+// orderedGroups returns a slice of *rules.Group from `groupsMap` which follows the order
+// mentioned by `groupOrderMap`. The returned slice starts with the groups that need to be in order,
+// followed by rest of the groups in undetermined order.
+func orderedGroups(groupsMap map[string]*rules.Group, groupOrderMap map[string]int) []*rules.Group {
+	var groups []*rules.Group
```

Comment:
Okay, this is too complex. Can we do this instead:

```go
groups := make([]*rules.Group, 0, len(groupsMaps))
for _, g := range groupsMap {
  groups = append(groups, g)
}

sort.Slice(groups, func(i, j int) bool {
  return groupOrderMap[groups[i].Name()] < groupOrderMap[groups[j].Name()]
})

return groups
```

Reply:
Didn't think of that, thanks! I have updated.